model/scratch/dffml_model_scratch/ridge.py

```diff
@@ -0,0 +1,219 @@
+        self.logger.debug(
+            "Predicted Value of {} {}:".format(
+                self.parent.config.predict.NAME, prediction
+            )
+        )
+        return prediction
+
+    async def squared_error(self, ys, yline):
+        return sum((ys - yline) ** 2)
+
+    async def coeff_of_deter(self, ys, regression_line):
+        y_mean_line = [np.mean(ys) for y in ys]
+        squared_error_mean = await self.squared_error(ys, y_mean_line)
+        squared_error_regression = await self.squared_error(
+            ys, regression_line
+        )
+        return 1 - (squared_error_regression / squared_error_mean)
+
+    async def best_fit_line(self):
+        self.logger.debug("Number of input repos: {}".format(len(self.xData)))
+        await self.fit(self.xData, self.yData)
+    
+    async def _preprocess_data(self, X, y, fit_intercept, normalize=False, copy=True,
+                     sample_weight=None, return_mean=False):
+
+        if isinstance(sample_weight, numbers.Number):
+            sample_weight = None
+        if copy:
+            X = X.copy(order='K')
+
+        y = np.asarray(y, dtype=X.dtype)
+
+        if fit_intercept:
+            X_offset = np.average(X, axis=0, weights=sample_weight)
+            X -= X_offset
+            if normalize:
+                X, X_scale = f_normalize(X, axis=0, copy=False,
+                                         return_norm=True)
+            else:
+                X_scale = np.ones(X.shape[1], dtype=X.dtype)
+            y_offset = np.average(y, axis=0, weights=sample_weight)
+            y = y - y_offset
+        else:
+            X_offset = np.zeros(X.shape[1], dtype=X.dtype)
+            X_scale = np.ones(X.shape[1], dtype=X.dtype)
+            if y.ndim == 1:
+                y_offset = X.dtype.type(0)
+            else:
+                y_offset = np.zeros(y.shape[1], dtype=X.dtype)
+
+        return X, y, X_offset, y_offset, X_scale
+
+
+    async def fit(self, X, y, sample_weight=None):
+        X = np.asarray(X)
+        y = np.array(y)
+        if len(X.shape) < 2:
+            X = X.reshape(-1, 1)
+        X, y, X_offset, y_offset, X_scale = await self._preprocess_data(
+                X, y, self.fit_intercept, self.normalize, self.copy_X,
+                sample_weight=sample_weight, return_mean=True)
+        params = {}
+        self.coef_, self.n_iter_ = await _ridge_regression(
+                X, y, alpha=self.alpha, sample_weight=sample_weight,
+                max_iter=self.max_iter, tol=self.tol,
+                random_state=self.random_state, return_n_iter=True,
+                return_intercept=False, check_input=False, **params)
+        await self._set_intercept(X_offset, y_offset, X_scale)
+        self.accuracy_value = await self.coeff_of_deter(y, \
+            np.dot(np.asarray(X, dtype=float), self.coef_.T)  + self.intercept_)
+        with open("weights.txt", 'w') as f:

```

Comment:
You'll want to save and load from `self.parent.config.directory` (+ `self._features_hash_` and whatever else might be needed to differentiate one trained model from another).

Typical use case of DFFML is from the command line, to make it easy to use, we don't require that users specify a directory. As a default, we save their models in directories / files  (depending on model implementation) and try to use the unique arguments to a model as input data to a hash function so that we get a filename that will work and map back to the same saved state whenever the same arguments are given.

Reply:
ok

Reply:
file_addr = self.parent.config.directory + "\\" + self._features_hash_ :added